Error norms

kkhemc opened the pull request:
Here are some additions to Kripke that allows it to compute error norms for the type "i" problems, i.e. no scattering.  You would run kripke with something like
kripke.exe --sigs 0,0,0 --zones 200,400,200 --quad 8:4 --niters 1 --compute_errors

If this is useful, we could add diagnostics comparing to the angle integrated total fluxes for the "type ii" problems.  The added code has the necessary benchmark data, we would just need to add the specific diagnostic calculations.

kkhemc committed:
added error norms for type i, no scattering, problems

kkhemc committed:
added missing ErrorNorms.cpp file to Kernel

Commit:
fixed a bug in the error normalization in parallel

ajkunen reviewed (changes requested):
This looks really good, just a few things would be great:
1. I'm getting a number of compiler warnings, mostly related to the need to convert some initializer lists to use double curly braces (for subobject initialization)
2. It would be nice if the --compute_errors flag was added to the --help output
3. It would be nice if there was some mention of it in the README file, with the note about running with sigs=0
4. This doesn't compile on GPU machines.  Two options: either make it work on the GPU (this may be difficult) or make it a optional feature that can be controlled via CMake (like -DENABLE_COMPUTE_ERRORS).  That way we can turn it off (and not compile it) for GPU builds

kkhemc committed:
1) fixed some compiler warnings. 2) added --compute_errors to the usage message. 3) added something about the error code to the README

kkhemc committed:
fixed issues with errors on gpus

kkhemc committed:
cleaned up some extraneous code

kkhemc commented:
I believe I have addressed all the issues in the review.  

I made some changes so that the errors work on gpu builds.  Basically I added made policies for ErrorNorms that are sequential when GPU builds are enabled.  I built and tested these on lassen.  

Please let me know if things are more corrections and/or additions to make.

ajkunen commented:
@kkhemc Thanks! I'll try and take a look at this over the weekend or next week.

brunner6 commented:
@ajkunen  and @rchen20 : This MR is really old.  If this is still worth merging, we should.  Otherwise we should close this.  @kkhemc : I'm sorry this has taken more than a few years to evaluate...
